### PR TITLE
Begin work on version 4.0 API changes

### PR DESCRIPTION
**Pull request by drmohundro**
This is to track the initial work on #84 - I'm not positive how many new features will make this release, but I'd definitely like to clean up some of the nil checks (whether by allowing defaults or some other means).

I'd welcome any additional PRs to this branch if others would like to see additional changes make it for the 4.0 release.

### CODE REVIEW

**Comment by norio-nomura** [comment]
FYI, we found that Swift 4.0 is introducing breaking changes that affects to SWXMLHash and filed at JIR [SR-4951](https://bugs.swift.org/browse/SR-4951).
I don't think that issue may be avoidable by using `-swift-version 3`.

**Comment by drmohundro** [comment]
@norio-nomura thanks for the heads up - I'll start watching that issue, but can plan to rename `Element` if necessary.

**Comment by norio-nomura** [comment]
> but can plan to rename Element if necessary.

Just an `Element`? or lowercase enums?

**Comment by drmohundro** [comment]
Yeah, I think lower case makes sense - I think I had intended to follow Swift conventions, but most of the things I'm seeing now are lowercase so that makes the most sense to me right now.

**Comment by norio-nomura** [comment]
I have a [branch](https://github.com/norio-nomura/SWXMLHash/tree/nn-lowercase-enums) that changes enums to lowercase for building with Swift 4.

**Comment by drmohundro** [comment]
Nice! I guess the only thing that would have to change before merging it into the 4.0 branch would be the `@available` `renamed` attributes. I'm happy to merge that into 4.0 here.

File: Source/SWXMLHash.swift [hunk 1]
```diff
@@ -444,27 +444,81 @@ public class IndexOps {
...
-    case XMLError(IndexingError)
+public enum XMLIndexer {
+    case element(XMLElement)
+    case list([XMLElement])
+    case stream(IndexOps)
+    case xmlError(IndexingError)
+
+// swiftlint:disable identifier_name
+    // unavailable
+#if !swift(>=3.2) // `Sequence.Element` is added on Swift 4.0 including with `-swift-version 3`
```
Comment (norio-nomura) [inline Source/SWXMLHash.swift]: This `#if` is no longer needed, since it is no longer `Sequence`.

Comment (norio-nomura) [inline Source/SWXMLHash.swift]: Fixed this in #138

**Pull request merged**